server/config/passport.ts

```diff
@@ -5,15 +5,32 @@ import type {Request} from 'express';
 
 import config from '../../config';
 import Users from '../models/Users';
+import {authAuthenticationSchema, authHTTPBasicAuthenticationSchema} from '../../shared/schema/api/auth';
+import {Credentials} from '../../shared/schema/Auth';
+import getAuthToken from '../util/authUtil';
 
 // Setup work and export for the JWT passport strategy.
 export default (passport: PassportStatic) => {
   const options = {
     jwtFromRequest: (req: Request) => {
       let token = null;
-
-      if (req && req.cookies) {
-        token = req.cookies.jwt;
+      let parsedResult = authAuthenticationSchema.safeParse(null);
+      let credentials: Required<Pick<Credentials, 'username' | 'password'>>;
+      switch (config.authMethod) {
+        case 'httpbasic':
+          parsedResult = authHTTPBasicAuthenticationSchema(req.header('authorization'));
+          if (!parsedResult.success) {
+            return token;
+          }
+
+          credentials = parsedResult.data;
+          token = getAuthToken(credentials.username);
```

Comment:
I wanted to rely only on the http basic credentials, but actully I tried another implementation before this final:
- the `jwtFromRequest` will remain as it is in master
- I added `passReqToCallback: true` to the passport options and the on https://github.com/jesec/flood/blob/master/server/config/passport.ts#L26 I will compare `jwtPayload.username` with http basic username and continue only if they match.
please, let me know this is the accepted solution so I can implement

Reply:
There is no need to modify this file. You only need to add logics to /verify and /authenticate endpoints. Simply parse the header to get the username and then reply with a generated token.

Reply:
and then for all the other endpoints you could send different credentials between jwt cookie and http basic.
that's really a mess, in my opinion.

Reply:
Other endpoints simply don't handle header. They only validate the cookie. Only selected endpoints handle alternative auth methods. This allows uniformity in auth enforcement and prevents errors in authentication system which can be disastrous.

Reply:
I totally disagree. It can be disastrous the contrary: mixing endpoints that check for authorization header and others that check only the cookie. This would not be unform at all.

Imagine a setup like mine where I have multiple upstream on different proxy path with different authentication credentials (one for flood, one for every other): the browser could cache different credetentials for one path and send again to the other path. this should lead to block the request to flood as unauthorized and force you to authorize again.

While the initial solution to rely only on the token generated by the authorization header username was not proper, both mechanism should be involved and validated: like the in the last commit I sent.

Reply:
No, we can't not check both header and cookie everywhere. And we may not grant token in endpoints other than /verify and /authenticate as that nullifies all anti-session-hijacking protections. 

If nested routes and conflicting cookie names are problems, we shall rename the current cookie from `jwt` to a more Flood-specific name.

Reply:
> No, we can't not check both header and cookie everywhere. And we may not grant token in endpoints other than /verify and /authenticate as that nullifies all anti-session-hijacking protections.

we keep granting token only in `/verify` and `/authenticate`
there's no nullification of any anti-session-hijacking protections.
if the auth is `httpbasic` on top of jwt token check there's another requirement: the http basic username should match the one in the jwt token. but the jwt token granting and validation was not changed.

what is the reason why we we cannot check both header and cookie?

Reply:
OK. That makes sense to me. Thinking about it, there is indeed a possibility of inconsistent header/cookie authentication state. 

I am just worry about the added complexity and non-uniformity which can be troublesome in security-related components.

Reply:
Makes sure this added logic is an added step. Don't replace other logics for simplicity and the peace of mind that this won't hurt security.

Reply:
> there is indeed a possibility of inconsistent header/cookie authentication state.
indeed: imagine the following setup (all behind a nginx proxy that forward to related upstream and different httpasswd by path):
- http://adomain.com/flood (users in httpasswd: `guest`, `aFloodUser`)
- http://adomain.com/grafana (users in httpasswd: `guest`, `admin`)

I login to the first as `aFloodUser`
I open a new tab for the second logging as `guest`
I would not receive a 401 from `/flood` but still the username will differ from the one in the jwt token


> Makes sure this added logic is an added step. Don't replace other logics for simplicity and the peace of mind that this won't hurt security.

sure, that's the whole point: inject new logic on in the switch branch for httpbasic. please validate during review that I didn't change the current behaviour in different/default branches

Reply:
Great. The logics look OK to me. However, the organization of functions needs some improvements in my opinion. May I push directly to the PR to tweak it a bit?

Reply:
great. let me push the last changes with frotend handling the prefilled registration and then feel free to push any tweak to the PR.

Reply:
I posted the patch: https://github.com/jesec/flood/pull/70#issuecomment-720114294. You can apply it by yourself. Or would you want me to apply it?